src/plugins/ui_actions/public/service/ui_actions_execution_service.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { uniqBy } from 'lodash';
+import { Action } from '../actions';
+import { BaseContext } from '../types';
+import { defer as createDefer, Defer } from '../../../kibana_utils/public';
+import { buildContextMenuForActions, openContextMenu } from '../context_menu';
+import { Trigger } from '../triggers';
+
+interface ExecuteActionTask {
+  action: Action;
+  context: BaseContext;
+  trigger: Trigger;
+  defer: Defer<void>;
+}
+
+export class UiActionsExecutionService {
+  private readonly batchingQueue: ExecuteActionTask[] = [];
+  private readonly pendingTasks = new Set<ExecuteActionTask>();
+
+  constructor() {}
+
+  async execute({
+    action,
+    context,
+    trigger,
+  }: {
+    action: Action<any>;
```

Comment:
Can you avoid the `any` here by using `Action`? Or maybe it should be something like:

```
execute<C extends BaseContext>({
    action,
    context,
    trigger,
  }: { action: Action<C>, context: C, trigger: Trigger }
```

Although perhaps UiActions type clean up should get its own isolated PR.